### DISCUSSION TITLE
Can't compile the open telemetry for C++14

### DISCUSSION THREAD

**Discussion by PeterStayPool** [open-telemetry/opentelemetry-cpp#2369] [Q&A] [answered]
Hi, All. 

We are currently evaluating the open telemetry for our C++ service. We used the following CMAKE options to compile and build the source code: We are building it in CentOS 7 (centos-release-scl-2-3.el7.centos.noarch)

cmake /builds/edge-data-services/opentelemetry-cmake-build/_deps/opentelemetry-src -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/builds/edge-data-services/opentelemetry-cmake-build/src/clang.linux-x64/Debug -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_CXX_STANDARD=14 -DBUILD_TESTING=OFF -DBUILD_SHARED_LIBS=OFF -DCMAKE_POSITION_INDEPENDENT_CODE=ON -DCMAKE_CXX_FLAGS_INIT=-fvisibility=hidden -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_CXX_FLAGS=-pthread
-- The C compiler identification is GNU 8.3.1
-- The CXX compiler identification is Clang 9.0.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /opt/rh/devtoolset-8/root/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/src/dev-tools/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building for architecture ARCH=x64
-- Performing Test check_cxx_compiler_flag_-Wno-type-limits
-- Performing Test check_cxx_compiler_flag_-Wno-type-limits - Success
-- Performing Test check_cxx_compiler_flag_-Wno-deprecated-declarations
-- Performing Test check_cxx_compiler_flag_-Wno-deprecated-declarations - Success
-- Performing Test check_cxx_compiler_flag_-Wno-unused-parameter
-- Performing Test check_cxx_compiler_flag_-Wno-unused-parameter - Success
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
Building with nostd types...
-- Configuring done
-- Generating done
-- Build files have been written to: /builds/edge-data-services/opentelemetry-cmake-build/.local/build

We are trying to link statically into our code and getting the following errors.

```
external/thirdparty_opentelemetry-cmake-build__linux/Release/include/opentelemetry/nostd/./internal/absl/types/../utility/../base/internal/../../meta/type_traits.h:625:37: error: unknown template name 'invoke_result'
struct result_of<F(Args...)> : std::invoke_result<F, Args...> {};
```

This error indicates that the linker is still looking for C++17 symbols. 
Can anyone tell us what's the right CMAKE options to build for C++14?

To give you the background, our toolchain does not support STL C++17 features, it only supports C++17 language features.
Currently, it only supports STL C++14.

I wonder whether we can build the open telemetry C++ for C++14 and with STL or STL C++14.

**Comment by marcalff** [comment, marked as answer]
Analysis as follows.

The error message seen at build time, by the compiler (not the linker), comes from:

```
#if __cplusplus >= 201703L
// std::result_of is deprecated (C++17) or removed (C++20)
template<typename> struct result_of;
template<typename F, typename... Args>
struct result_of<F(Args...)> : std::invoke_result<F, Args...> {};
#else
template<typename F> using result_of = std::result_of<F>;
#endif
```

`__cplusplus >= 201703L` is testing that the compiler supports C++17 language features

`std::invoke_result` is part of the C++17 STL, so in this case, in an environment with a C++17 compiler / C++ 14 STL, this fragment of code will fail.

A work around is to build in C++14 instead, not sure if this is compatible with your requirements.

To do so, please note that:

* building opentelemetry-cpp should use `-DCMAKE_CXX_STANDARD=14`, which seems to be the case
* building the application should **_also_** use `-DCMAKE_CXX_STANDARD=14`, which seems to be missing since the compiler still sees `__cplusplus > 201703L`

About building opentelemetry-cpp with the C++14 STL only, this is supported starting with opentelemetry-cpp version 1.12.0 released today.

See in particular:

* https://github.com/open-telemetry/opentelemetry-cpp/releases/tag/v1.12.0
* https://github.com/open-telemetry/opentelemetry-cpp/pull/2304
